Automattic/woocommerce-payments
# Webhook Reliability: Fetch failed events and process them

Fixes #3813

#### Changes proposed in this Pull Request

Add a new service to achieve by doing these things: 

- Hook into `woocommerce_payments_account_refreshed` and check the relevant flag, and decide whether to schedule a job `wcpay_webhook_fetch_events`. Relevant convo p1645601664622659/1645086273.563649-slack-C01BPL3ALGP
- `wcpay_webhook_fetch_events` jobs fetches events from the server, then do two things: 
     - Decide whether to schedule another `wcpay_webhook_fetch_events`.
     - Save data and schedule a job for each event.
- `wcpay_webhook_process_event` jobs simply retrieve data and dispatch to `WC_Payments_Webhook_Processing_Service` (see this PR https://github.com/Automattic/woocommerce-payments/pull/3852)









### Testing instructions

Requirement: this server PR 1633-gh-Automattic/woocommerce-payments-server
ActionScheduler management pending page: wp-admin/tools.php?page=action-scheduler&status=pending



#### Test case: properly schedule `wcpay_webhook_fetch_events`

**TODOs:** 

- [x] Wait for the server PR to implement this. Otherwise, you can apply this diff for testing 2bd57-pb/#diff - change true/false to see the difference. 
- [x] Update this value `CONTINUOUS_FETCH_FLAG_ACCOUNT_DATA` based on the server PR.

Steps: 

- Follow the server PR, and ensure that your site has some failed not-yet-processed webhook events.
- Go to ActionScheduler management page and run `wcpay_refresh_account_cache`, which forces to retrieve new account data. 
- Expect: a new job `wcpay_webhook_fetch_events` is scheduled. 

#### Test case: run `wcpay_webhook_fetch_events`

- Ensure that your site has some not-yet-processed failed webhook events.
- Go to ActionScheduler management page and run `wcpay_webhook_fetch_events`.
- Expect: failed events are retrieved from the server, and new jobs `wcpay_webhook_process_event` are scheduled with the `event_id` parameter. 
- Try adding more than 10 not-yet-processed failed webhook events. 
- Go to ActionScheduler management page and run `wcpay_webhook_fetch_events`.
- Expect: similar above plus that another `wcpay_webhook_fetch_events` is scheduled (This is to test the `has_more` flag).
- Expect: Go to wp-content/uploads/wc-logs (if you use Docker, it's under the `docker/wordpress` folder), find the log file for today, and see the relevant log entries. Tip: use `event_id` for search.

#### Test case: run `wcpay_webhook_process_event`. 

- Go to ActionScheduler management page and run a `wcpay_webhook_process_event` job.
- Expect: Also check in the log file and see the relevant log entries for `event_id` attached to the job you've just run. 



-------------------


**Post merge**
